src/QMCWaveFunctions/OptimizableObject.h

```diff
@@ -43,31 +43,35 @@ class OptimizableObject
   /** check in variational parameters to the global list of parameters used by the optimizer.
    * @param active a super set of optimizable variables
    *
-   * This is a query function and should never be implemented as a feature blocker.
-   * If an OptimizableObject derived class doesn't support optimization, use the base class fallback.
+   * The existing checkInVariables implementation in WFC/SPO/.. are inclusive and it calls checkInVariables of its members
+   * class A: public SPOSet {}
+   * class B: public WFC
+   * {
+   *   A objA;
+   *   checkInVariables() { objA.checkInVariables(); }
+   * };
+   *
+   * With OptimizableObject,
+   * class A: public OptimizableObject {}
+   * class B: public OptimizableObject
+   * {
+   *   A objA;
+   *   checkInVariablesExclusive() { // should not call objA.checkInVariablesExclusive(); }
+   * };
+   * A vector will be created by calling extractOptimizableObjects().
+   * All the checkInVariables() will be called through this vector and thus
+   * checkInVariablesExclusive implementation should only handle non-OptimizableObject members.
```

Comment:
So to me the use of inclusive and exclusive here is not very clear.  I immediately thought of it terms of thread when I saw the Exclusive on this method. i.e. only call if you are certain you will be the only caller. Exclusive/Inclusive I might also have the sense of a range that does or doesn't include the endpoints or an operation that checks in any variable not already checked in (XOR/OR).

The difference here is whether the call traverses the implicit tree of optimizables below the object called with checkInVariables or not. So its easier for me to think of it as a "Traversing" checkInVariables (the legacy call) or a NonTraversing the new `checkInVariablesExclusive` call. Or as a recursive vs. nonrecursive call.

Reply:
It does seem clear in the new design the traversal occurs with extractOptimizableObjects so the checkInVariables must be non traversing.

I don't think this is a bad design especially since the "tree" is adhoc and each "node" used to class handle calling its leaves/branches in an adhoc way.  

Now that "tree" traversal is in classes extract method and the checkIn/reset/whatever are in methods that just do that which is better.

Reply:
I was taking the notion from inclusive/exclusive timer.